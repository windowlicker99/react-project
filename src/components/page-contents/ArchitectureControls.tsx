import React, { FC, useEffect, useState } from 'react';
import { Btn } from '@components/common/form-controls/Btn';
import { Badge } from '@components/common/Badge';
import { Legend } from '@components/common/Legend';
import { PageFilter } from '@components/page-contents/PageFilter';
import { CircleChart } from '@components/common/CircleChart';
import { domainFilters } from '@constants/Filters';
import { IVehicle } from '@interfaces/interfaces';
import { EBadgeTypes, ERiskLevels } from '@interfaces/enums';
import { initialVehicle } from '@constants/initialData';

interface IArchitectureControls {
  vehicle: IVehicle;
  onOpenDomainModalClick: () => void;
  onOpenHistoryView: () => void;
}

export const ArchitectureControls: FC<IArchitectureControls> = ({
  vehicle,
  onOpenDomainModalClick,
  onOpenHistoryView,
}) => {
  const [displayedVehicle, setDisplayedVehicle] = useState<IVehicle>(initialVehicle);
  const [isOpenFilters, setIsOpenFilters] = useState<boolean>(false);

  const onShowFiltersClick = () => {
    setIsOpenFilters(!isOpenFilters);
  };

  const refreshFilters = () => {
    // temporary solution. Will be removed with adding filters refresh logic
    console.log('refresh filters');
  };

  useEffect(() => {
    if (!vehicle) {
      setDisplayedVehicle(initialVehicle);
      return;
    }
    setDisplayedVehicle(vehicle);
  }, [vehicle]);

  return (
    <div className="box page-controls">
      <div className="page-controls-content">
        <div className="page-controls-item">
          <div className="mr-15">
            <Btn
              disabledBtn={!vehicle}
              btnStatus="btn-lg-primary"
              btnIcon="icon-plus"
              btnText="Create Domain"
              onClick={onOpenDomainModalClick}
            />
          </div>
          <div className="mr-15">
            <Btn btnStatus="btn-lg-outline" btnText="View History" onClick={onOpenHistoryView} />
          </div>
          <div className="mr-15">
            <Btn btnStatus="btn-xs-primary" btnIcon="icon-filter-blue" onClick={onShowFiltersClick} />
          </div>
          <div className="mr-15">
            <Btn btnStatus="btn-xs-secondary" btnIcon="icon-refresh" onClick={refreshFilters} />
          </div>
        </div>
        <div className="page-controls-item">
          <div className="page-controls-architecture">
            <div className="page-controls-data">
              {vehicle && (
                <div className="page-controls-info">
                  <span className="page-controls-name">{displayedVehicle.brand?.name}</span>
                  {displayedVehicle.model}
                </div>
              )}
            </div>
            <div className="page-controls-data">
              Risk Level
              <Badge
                badgeType={EBadgeTypes[displayedVehicle.riskLevel]}
                badgeText={ERiskLevels[displayedVehicle.riskLevel]}
              />
            </div>
            <div className="page-controls-data">
              Progress
              <div className="chart">
                <CircleChart percentage={displayedVehicle.raProgress} />
              </div>
            </div>
            <div className="page-controls-data">
              <Legend isDisabled={!vehicle} legendType="Eth" legendName="Ethernet" />
              <Legend isDisabled={!vehicle} legendType="CAN" legendName="CAN" />
              <Legend isDisabled={!vehicle} legendType="LIN" legendName="LIN" />
            </div>
          </div>
        </div>
      </div>
      <PageFilter filters={domainFilters} isOpen={isOpenFilters} close={onShowFiltersClick} />
    </div>
  );
};
