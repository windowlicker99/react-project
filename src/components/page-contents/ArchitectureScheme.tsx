import React, { FC, MouseEventHandler, useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  ConnectorModel,
  DataBinding,
  DiagramComponent,
  Inject,
  NodeModel,
  SnapConstraints,
  SnapSettingsModel,
} from '@syncfusion/ej2-react-diagrams';
import { TooltipComponent } from '@syncfusion/ej2-react-popups';
import { createNodes } from '@utils/createNodes';
import { createConnectors } from '@utils/createConnectors';
import { calculateSchemeOffsetAndScale } from '@utils/calculateSchemeOffsetAndScale';
import {
  IArchitectureDomain,
  IDiagramClickEventArgs,
  ISelectionChangeArgs,
  IDiagramTooltip,
  IElementAddInfo,
} from '@interfaces/ArchitectureScheme/interfaces';
import { setCurrentElement } from '@store/slices/elementSlice';
import { selectedItemsSettings, tooltipAnimationParams } from '@constants/ArchitectureScheme/parameters';
import { createDiagramTooltip } from '@utils/createDiagramTooltip';
import { selectCurrentVehicle } from '@store/selectors/vehicle';
import { createGateway } from '@utils/createGateway';
import { PageMock } from './PageMock';
import { getAddedNodes } from '@/utils/getAddedNodes';
import { EElementPropName } from '@/interfaces/ArchitectureScheme/enums';
import { TDiagramEventElement } from '@/interfaces/ArchitectureScheme/types';
import { IECUasGateway, IGatewayResponse } from '@/interfaces/interfaces';

interface IArchitectureSchemeProps {
  onECUCreateClick: () => void;
  schemeData: IArchitectureDomain[];
  gatewayInfo: IGatewayResponse;
}

const cloneDeep = require('lodash.clonedeep');

export const ArchitectureScheme: FC<IArchitectureSchemeProps> = ({ schemeData, gatewayInfo, onECUCreateClick }) => {
  const dispatch = useDispatch();
  const currentVehicle = useSelector(selectCurrentVehicle);
  const [gatewayNode, setGatewayNode] = useState<NodeModel>(null);
  const [nodes, setNodes] = useState<NodeModel[]>([]);
  const [connectors, setConnectors] = useState<ConnectorModel[]>([]);
  const [tooltip, setTooltip] = useState<IDiagramTooltip>({ text: '' });
  const [selectedConnectors, setSelectedConnectors] = useState([]);
  const clonedConnectors = cloneDeep(connectors);
  const [addedNodes, setAddedNodes] = useState<NodeModel[]>([]);
  let diagramInstance: DiagramComponent = null;

  if (!currentVehicle) {
    return <PageMock />;
  }

  const changeSelectedConnectorStyle = (element: TDiagramEventElement, width: number) => {
    if (!element?.addInfo || element.addInfo.relative) {
      return;
    }
    connectors.forEach((connector, i) => {
      const domainId = `domain${element.addInfo.element?.id}_to`;
      if (connector.id.includes(domainId)) {
        clonedConnectors[i].style.strokeWidth = width;
      }
    });
  };

  const getNodeHTML = (element: TDiagramEventElement) =>
    document.querySelector(`#${element.id}_content_html_element > div > div`);

  const changeSelectedStyle = (element: TDiagramEventElement, isNew: boolean) => {
    if (element?.data === EElementPropName.connector) {
      const width = isNew ? 2 : 1;
      changeSelectedConnectorStyle(element, width);
    }
    if (element?.data === EElementPropName.node) {
      const node = getNodeHTML(element);
      if (!node) {
        return;
      }
      if (isNew) {
        node.classList.add('selected');
        return;
      }
      node.classList.remove('selected');
    }
  };

  const manageStylesOfSelectedItems = (args: ISelectionChangeArgs): void => {
    const {
      newValue: [newValue],
      oldValue: [oldValue],
    } = args;

    changeSelectedStyle(newValue, true);
    changeSelectedStyle(oldValue, false);

    // update connectors
    setConnectors(clonedConnectors);
  };

  const changeColorConnection = (args: ISelectionChangeArgs) => {
    const {
      newValue: [element],
    } = args;
    const relativeConnectors: number[] = [];

    connectors.forEach((connector, index) => {
      const addInfo = connector.addInfo as IElementAddInfo;

      if (
        (connector.sourceID === element?.id && addInfo?.relative) ||
        (element?.id === connector.targetID && addInfo?.relative)
      ) {
        relativeConnectors.push(index);
      }
    });

    relativeConnectors.forEach((i) => {
      clonedConnectors[i].style = {
        strokeColor:
          clonedConnectors[i].style.strokeColor === '#BEBEBE' ? clonedConnectors[i].addInfo?.relative.color : '#BEBEBE',
      };
    });

    selectedConnectors.forEach((i) => {
      clonedConnectors[i].style = {
        strokeColor:
          clonedConnectors[i].style.strokeColor === '#BEBEBE' ? clonedConnectors[i].addInfo?.relative.color : '#BEBEBE',
      };
    });

    setSelectedConnectors(relativeConnectors);

    setConnectors(clonedConnectors);
  };

  const getGatewayNode = async () => {
    if (!gatewayInfo) {
      setGatewayNode(null);
      return;
    }
    const updateGatewayNode = createGateway(gatewayInfo);

    setGatewayNode(updateGatewayNode);
  };

  const getNodes = () => {
    const schemeNodes = createNodes(schemeData, gatewayNode);
    if (nodes.length) {
      const added = getAddedNodes(nodes, schemeNodes);
      setAddedNodes(added);
    }
    setNodes(gatewayNode ? [gatewayNode, ...schemeNodes] : schemeNodes);
    const gateway = gatewayInfo?.gateway as IECUasGateway;
    const schemeConnectors = createConnectors(schemeData, gatewayNode?.height, gateway?.color);
    setConnectors(schemeConnectors);
  };

  useEffect(() => {
    getGatewayNode();
  }, [gatewayInfo]);

  useEffect(() => {
    if (schemeData) {
      getNodes();
    }
  }, [schemeData, gatewayNode]);

  const snapSettings: SnapSettingsModel = {
    constraints: SnapConstraints.None,
  };

  const onShowElementDetails = (args: ISelectionChangeArgs) => {
    const {
      newValue: [element],
    } = args;
    if (element?.addInfo) {
      const addInfo = element.addInfo as IElementAddInfo;
      if (addInfo.element) {
        dispatch(setCurrentElement(addInfo.element));
      }
    } else {
      dispatch(setCurrentElement(null));
    }
  };

  const onSelectElement = (args: ISelectionChangeArgs) => {
    manageStylesOfSelectedItems(args);
    if (!args.newValue?.length) {
      dispatch(setCurrentElement(null));
    } else {
      onShowElementDetails(args);
      changeColorConnection(args);
    }
  };

  const onDiagramMisclick = ({ actualObject }: IDiagramClickEventArgs) => {
    if (!actualObject) {
      diagramInstance.select([]);
    }
  };

  const onMouseOver = ({ actualObject }: IDiagramClickEventArgs) => {
    if (actualObject?.addInfo) {
      const { addInfo } = actualObject;

      if (addInfo.tooltip) {
        setTooltip(addInfo.tooltip);
      }
    }
  };

  const onMouseLeave = () => {
    setTooltip({ text: '' });
  };

  const onOpenModal: MouseEventHandler = (e) => {
    const { target } = e;
    if (!(target instanceof HTMLElement)) {
      return;
    }
    if (target.dataset.createEcu) {
      onECUCreateClick();
    }
  };

  useEffect(() => {
    // to update nodes in the diagram correctly
    if (diagramInstance) {
      if (addedNodes.length) {
        diagramInstance.clearSelection(); // remove selection from add button
        diagramInstance.select(addedNodes); // show details of new element
      }
      nodes.forEach((node) => {
        diagramInstance.add(node);
      });
    }
  }, [nodes]);

  useEffect(() => {
    if (diagramInstance) {
      // preserves the interactivity of the connectors
      diagramInstance.connectors = [...connectors];

      // fixes the issue of disappearing connectors
      connectors.forEach((connector) => {
        diagramInstance.add(connector);
      });
    }
  }, [connectors]);

  useEffect(
    () => () => {
      diagramInstance = null;
    },
    []
  );

  if (nodes?.length) {
    return (
      <TooltipComponent
        isSticky
        closeDelay={300}
        content={createDiagramTooltip(tooltip)}
        mouseTrail
        showTipPointer
        position="BottomCenter"
        offsetY={10}
        animation={tooltipAnimationParams}
      >
        <DiagramComponent
          ref={(diagram) => {
            diagramInstance = diagram;
          }}
          selectedItems={selectedItemsSettings}
          id="diagram"
          width="100%"
          height="100%"
          snapSettings={snapSettings}
          scrollSettings={calculateSchemeOffsetAndScale(schemeData?.length)}
          mouseOver={onMouseOver}
          mouseLeave={onMouseLeave}
          selectionChange={onSelectElement}
          click={onDiagramMisclick}
          onClick={onOpenModal}
        >
          <Inject services={[DataBinding]} />
        </DiagramComponent>
      </TooltipComponent>
    );
  }

  return null;
};
