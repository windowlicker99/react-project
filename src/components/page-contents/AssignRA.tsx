import React, { FC, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { useFormik } from 'formik';
import { ModalCenter } from '@components/modules/ModalCenter';
import { AssignTicketModalContent } from '@components/modal-content/AssignTicketModalContent';
import { selectCurrentVehicle } from '@store/selectors/vehicle';
import { getUsers } from '@utils/getUsers';
import { convertUsersToOptions } from '@utils/converters';
import { getElement } from '@utils/getElement';
import { createTicket } from '@utils/createTicket';
import { ICurrentElement, IOption } from '@interfaces/interfaces';
import { TElement } from '@interfaces/types';
import { EAsideActualElement, ETicketsTypes } from '@interfaces/enums';

interface IAssignRA {
  element: ICurrentElement;
  isOpen: boolean;
  onClose: () => void;
}

export const AssignRA: FC<IAssignRA> = ({ element, isOpen, onClose }) => {
  const defaultOption = { label: 'Not Assigned', value: 'NaN' };
  const [currentElement, setCurrentElement] = useState<TElement>(null);
  const [usersOptions, setUsersOptions] = useState<IOption[]>([defaultOption]);

  const vehicleId = useSelector(selectCurrentVehicle);

  const getUsersOptions = async () => {
    const users = await getUsers();
    const options = convertUsersToOptions(users);
    setUsersOptions([defaultOption, ...options]);
  };

  const getCurrentElement = async () => {
    const newElement = await getElement(element);
    setCurrentElement(newElement);
  };

  useEffect(() => {
    if (element) {
      getCurrentElement();
    }
  }, [element]);

  useEffect(() => {
    getUsersOptions();
  }, []);

  const formik = useFormik({
    initialValues: { user: defaultOption, message: '' },
    onSubmit: async ({ user: { value: userId }, message }, { resetForm }) => {
      try {
        await createTicket({
          name: `Risk Assessment for ${currentElement?.name}`,
          type: ETicketsTypes.riskAssessment,
          elementName: currentElement.name,
          elementType: EAsideActualElement[element.type],
          elementId: element.id,
          userId,
          vehicleId,
          message,
        });
      } catch (error) {
        console.log(error);
      } finally {
        resetForm();
        onClose();
      }
    },
  });

  const handleSelectChange = (name: string, value: IOption) => formik.setFieldValue(name, value);

  const onApplyRAModal = () => {
    formik.submitForm();
  };

  const onCloseRAModal = () => {
    formik.resetForm();
    onClose();
  };

  return (
    <ModalCenter
      title="Assign Ticket"
      subtitle={`Risk Assessment for ${currentElement?.name}`}
      isOpen={isOpen}
      disabledApplyBtn={!formik.values.user || formik.values.user.value === defaultOption.value}
      onApply={onApplyRAModal}
      onClose={onCloseRAModal}
    >
      <AssignTicketModalContent
        values={formik.values}
        options={usersOptions}
        onChange={formik.handleChange}
        onSelectChange={handleSelectChange}
      />
    </ModalCenter>
  );
};
