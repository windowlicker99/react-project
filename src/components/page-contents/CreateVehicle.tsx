import React, { useState, FC } from 'react';
import { Form, Formik, FormikProps } from 'formik';
import { ModalAside } from '@components/modules/ModalAside';
import { ModalFooter } from '@components/modal-content/modal-aside/ModalFooter';
import { ModalHeader } from '@components/modal-content/modal-aside/ModalHeader';
import { ResultContent } from '@components/modal-content/ResultContent';
import { VehicleCreateInfoContent } from '@components/modal-content/VehicleCreateInfoContent';
import { VehicleCreateDatesContent } from '@components/modal-content/VehicleCreateDatesContent';
import { getFieldsValues } from '@utils/getFieldsValues';
import { gatewayInfo, vehicleCreationDates, vehicleCreationElements } from '@constants/formsElements';
import { createVehicle } from '@utils/createVehicle';
import { convertVehicleFormValuesToPostData } from '@utils/converters';

import { vehicleBasicInfoValidationScheme, vehicleMilestoneDatesValidationScheme } from '@utils/validations';
import { IGatewayFormValues, IVehicleFormValues } from '@/interfaces/interfaces';

interface ICreateVehicle {
  isOpen: boolean;
  onClose: () => void;
}

type TPages = 1 | 2;

export const CreateVehicle: FC<ICreateVehicle> = ({ isOpen, onClose }) => {
  const [step, setStep] = useState<TPages>(1);
  const initialValues = {
    ...getFieldsValues(Object.values(vehicleCreationElements)),
    gateway: getFieldsValues(Object.values(gatewayInfo)) as IGatewayFormValues,
    ...getFieldsValues(Object.values(vehicleCreationDates)),
  };

  const [successfullyAdded, setSuccessfullyAdded] = useState<string>('');

  const onSubmit = async (values: IVehicleFormValues) => {
    try {
      const vehicle = convertVehicleFormValuesToPostData(values);
      await createVehicle(vehicle);
      setSuccessfullyAdded(`${values.brand?.label} | ${values.model} | ${values.platform} was created successfully!`);
    } catch (error) {
      console.log(error);
    }
  };

  const onConfirmClick = () => {
    setSuccessfullyAdded('');
    setStep(1);
    onClose();
  };

  const onPrevPageClick = () => setStep(1);
  const onNextPageClick = () => setStep(2);

  const onCloseModal = () => {
    setStep(1);
    onClose();
  };

  const pages = {
    1: {
      component: <VehicleCreateInfoContent />,
      header: {
        title: 'Create Vehicle',
        subtitle: 'Basic Information',
      },
      footer: {
        applyBtn: 'Next',
      },
      validationScheme: vehicleBasicInfoValidationScheme,
      onSubmit: onNextPageClick,
    },
    2: {
      component: <VehicleCreateDatesContent />,
      header: {
        title: 'Create Vehicle',
        subtitle: 'Milestone Dates',
      },
      footer: {
        applyBtn: 'Complete Creation',
        onBack: onPrevPageClick,
      },
      validationScheme: vehicleMilestoneDatesValidationScheme,
      onSubmit,
    },
  };

  return successfullyAdded ? (
    <ModalAside isOpen={isOpen}>
      <ResultContent title="Success" content={successfullyAdded} onConfirm={onConfirmClick} />
    </ModalAside>
  ) : (
    <Formik
      initialValues={initialValues}
      validationSchema={pages[step].validationScheme}
      onSubmit={pages[step].onSubmit}
    >
      {({ submitForm, resetForm, isValid, setTouched }: FormikProps<Partial<IVehicleFormValues>>) => (
        <ModalAside
          isOpen={isOpen}
          header={
            <ModalHeader
              step={step}
              maxStep={2}
              {...pages[step].header}
              onClose={() => {
                resetForm();
                onCloseModal();
              }}
            />
          }
          footer={
            <ModalFooter
              {...pages[step].footer}
              onApply={() => {
                submitForm();
                if (isValid) {
                  setTouched({});
                }
              }}
            />
          }
        >
          <Form> {pages[step].component}</Form>
        </ModalAside>
      )}
    </Formik>
  );
};
