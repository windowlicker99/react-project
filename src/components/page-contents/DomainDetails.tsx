import React, { useState, FC } from 'react';
import { useFormik } from 'formik';
import { useSelector } from 'react-redux';
import { ModalAside } from '@components/modules/ModalAside';
import { DomainDetailsContent } from '@components/modal-content/DomainDetailsContent';
import { ModalFooter } from '@components/modal-content/modal-aside/ModalFooter';
import { ModalHeader } from '@components/modal-content/modal-aside/ModalHeader';
import { selectUser } from '@store/selectors/auth';
import { getFieldsValues } from '@utils/getFieldsValues';
import { domainValidationSchema } from '@utils/validations';
import { addNewDomain } from '@utils/addNewDomain';
import { domainCreationElements } from '@constants/formsElements';
import { IOption, IVehicle } from '@interfaces/interfaces';
import { EElementCreationTabs } from '@/interfaces/enums';
import { DomainChooseContent } from '../modal-content/DomainChooseCollection';
import { elementDetailsTabs } from '@/constants/tabsLists';
import { ResultContent } from '../modal-content/ResultContent';
import { backgroundColors } from '@/constants/color';

interface IDomainDetails {
  vehicle: IVehicle;
  isOpenDomainModal: boolean;
  onCloseDomainModalClick: () => void;
}

export const DomainDetails: FC<IDomainDetails> = ({ vehicle, isOpenDomainModal, onCloseDomainModalClick }) => {
  const initialValues = getFieldsValues(Object.values(domainCreationElements));
  const [successfullyAdded, setSuccessfullyAdded] = useState<string>('');
  const [detailsForm, setDomainDetailsForm] = useState<string>(EElementCreationTabs.create);

  const user = useSelector(selectUser);

  const formik = useFormik({
    initialValues,
    validationSchema: domainValidationSchema,
    onSubmit: async ({ name, domainUsage, connectivity, color, communication: selectedCommunication }) => {
      try {
        const communication: string[] = selectedCommunication.map(({ label }: IOption) => label);
        const onlineConnectivity = connectivity ? 'Online' : 'Offline';
        await addNewDomain({
          name,
          domainUsage,
          onlineConnectivity,
          communication,
          color,
          background: backgroundColors[color as keyof typeof backgroundColors],
          authorId: user.id,
          vehicleId: vehicle.id,
        });
        setSuccessfullyAdded(
          `${name} was successfully added to ${vehicle.brand?.name} | ${vehicle.model} | ${vehicle.platform} architecture scheme!`
        );
      } catch (error) {
        console.log(error);
        onCloseDomainModalClick();
        setSuccessfullyAdded('');
      }
    },
  });

  const handleSelectChange = (name: string, value: IOption) => formik.setFieldValue(name, value);

  const onApplyModal = () => {
    formik.validateForm();
    formik.submitForm();
  };

  const onCloseModal = () => {
    formik.resetForm();
    onCloseDomainModalClick();
  };

  const onConfirmClick = () => {
    setSuccessfullyAdded('');
    onCloseModal();
  };

  return successfullyAdded ? (
    <ModalAside isOpen={isOpenDomainModal}>
      <ResultContent title="Success" content={successfullyAdded} onConfirm={onConfirmClick} />
    </ModalAside>
  ) : (
    <ModalAside
      isOpen={isOpenDomainModal}
      header={
        <ModalHeader
          title="Domain Details"
          activeTab={detailsForm}
          tabsList={elementDetailsTabs}
          onClose={onCloseModal}
          onChooseTabClick={setDomainDetailsForm}
        />
      }
      footer={
        <ModalFooter
          applyBtn="Create"
          onApply={onApplyModal}
          showDuplicateSwitch={detailsForm === EElementCreationTabs.choose}
        />
      }
    >
      {detailsForm === EElementCreationTabs.create ? (
        <DomainDetailsContent {...formik} handleSelectChange={handleSelectChange} />
      ) : (
        <DomainChooseContent />
      )}
    </ModalAside>
  );
};
