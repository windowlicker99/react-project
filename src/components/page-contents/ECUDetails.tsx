import React, { useState, FC } from 'react';
import { Form, Formik, FormikProps } from 'formik';
import { ModalAside } from '@components/modules/ModalAside';
import { ModalFooter } from '@components/modal-content/modal-aside/ModalFooter';
import { ModalHeader } from '@components/modal-content/modal-aside/ModalHeader';
import { ECUCreateNewStep1 } from '@components/modal-content/ECUCreateNewStep1';
import { ECUCreateNewStep2 } from '@components/modal-content/ECUCreateNewStep2';
import { ResultContent } from '@components/modal-content/ResultContent';
import { DomainChooseContent } from '@components/modal-content/DomainChooseCollection';
import { getFieldsValues } from '@utils/getFieldsValues';
import { addNewECU } from '@utils/addNewECU';
import { convertArrayToString, convertECUFormValuesToPostData } from '@utils/converters';
import { ECUCreation } from '@constants/formsElements';
import { elementDetailsTabs } from '@constants/tabsLists';
import { IECUFormValues, IOptionsList, IVehicle } from '@interfaces/interfaces';
import { EElementCreationTabs } from '@interfaces/enums';
import { ECUDetailsValidationScheme, ECUQuestionnaireValidationScheme } from '@utils/validations';

interface IECUDetails {
  vehicle: IVehicle;
  isOpen: boolean;
  onClose: () => void;
}

type TPages = 1 | 2;

export const ECUDetails: FC<IECUDetails> = ({ vehicle, isOpen, onClose }) => {
  const numberOfSteps = 2;
  const initialValues = {
    ...getFieldsValues(Object.values(ECUCreation)),
  };
  const [step, setStep] = useState<TPages>(1);
  const [successfullyAdded, setSuccessfullyAdded] = useState<string>('');
  const [detailsForm, setDomainDetailsForm] = useState<string>(EElementCreationTabs.create);

  const onPrevPageClick = () => setStep(1);
  const onNextPageClick = () => setStep(2);

  const onSubmit = async (values: IECUFormValues) => {
    try {
      const formData = convertECUFormValuesToPostData(values);
      const data = { ...formData, vehicleId: vehicle?.id };

      await addNewECU(data);
      const domainsOptions = values.domains.map(({ id }: IOptionsList) => ({ name: id?.label }));
      const domains = convertArrayToString(domainsOptions);

      setSuccessfullyAdded(
        `${data.name} was successfully added to ${domains} within ${vehicle.brand.name} | ${vehicle.model} | ${vehicle.platform} architecture scheme!`
      );
    } catch (error) {
      console.log(error);
    }
  };

  const pages = {
    1: {
      component: <ECUCreateNewStep1 vehicleId={vehicle?.id} updateContent={isOpen} />,
      header: {
        title: 'ECU Details',
      },
      footer: {
        applyBtn: 'Next',
      },
      validationScheme: ECUDetailsValidationScheme,
      onSubmit: onNextPageClick,
    },
    2: {
      component: <ECUCreateNewStep2 vehicleId={vehicle?.id} updateContent={isOpen} />,
      header: {
        title: 'ECU Risk Level Questionnaire',
      },
      footer: {
        applyBtn: 'Complete Creation',
        onBack: onPrevPageClick,
      },
      validationScheme: ECUQuestionnaireValidationScheme,
      onSubmit,
    },
  };

  const onCloseModal = ({ resetForm }: Partial<FormikProps<IECUFormValues>>) => {
    onClose();
    setStep(1);
    resetForm();
  };

  const onConfirmClick = () => {
    setStep(1);
    setSuccessfullyAdded('');
    onClose();
  };

  return successfullyAdded ? (
    <ModalAside isOpen={isOpen}>
      <ResultContent title="Success" content={successfullyAdded} onConfirm={onConfirmClick} />
    </ModalAside>
  ) : (
    <Formik
      initialValues={initialValues}
      onSubmit={pages[step].onSubmit}
      validationSchema={pages[step].validationScheme}
    >
      {({ submitForm, isValid, setTouched, ...props }: FormikProps<IECUFormValues>) => (
        <ModalAside
          isOpen={isOpen}
          header={
            <ModalHeader
              {...pages[step].header}
              step={step}
              maxStep={numberOfSteps}
              activeTab={detailsForm}
              tabsList={elementDetailsTabs}
              onClose={() => onCloseModal(props)}
              onChooseTabClick={setDomainDetailsForm}
            />
          }
          footer={
            <ModalFooter
              onApply={() => {
                submitForm();
                if (isValid) {
                  setTouched({});
                }
              }}
              {...pages[step].footer}
              showDuplicateSwitch={detailsForm === EElementCreationTabs.choose}
            />
          }
        >
          <Form>{detailsForm === EElementCreationTabs.create ? pages[step].component : <DomainChooseContent />}</Form>
        </ModalAside>
      )}
    </Formik>
  );
};
