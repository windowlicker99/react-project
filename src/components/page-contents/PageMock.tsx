import React, { FC } from 'react';

export const PageMock: FC = () => (
  <div
    style={{
      width: '100%',
      height: '100%',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      backgroundImage: `url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='190' height='151' viewBox='0 0 190 151'%3E%3Cg fill='none' fill-rule='evenodd'%3E%3Cpath d='M0 0h190v151H0z'/%3E%3Cg%3E%3Cellipse fill='%23E5F0FF' opacity='.504' cx='110.999' cy='64.93' rx='64.783' ry='64.93'/%3E%3Cpath d='m34.192 149.724-3.02-75.786a1.257 1.257 0 0 1 .33-.9l13.112-14.246c.238-.258.573-.405.925-.405h23.073c.322 0 .632.123.866.345l14.236 13.522c.233.222.543.346.866.346H95.24l-61.049 77.124z' fill='%236CA8FE'/%3E%3Cpath d='M36.356 140.015 34.42 90.57a1.257 1.257 0 0 1 1.256-1.306h17.788l-17.108 50.75z' fill='%23BBD7FE'/%3E%3Cpath d='m37.542 131.415 1.438-51.88a1.257 1.257 0 0 1 1.256-1.222h16.119l-18.813 53.102z' fill='%23E5F0FF'/%3E%3Cpath d='m37.373 133.723 6.873-61.171a1.257 1.257 0 0 1 1.395-1.108l43.213 5.07-51.481 57.209z' fill='%23FCFCFC'/%3E%3Cpath d='M99 66.334h51.205a1.257 1.257 0 0 1 1.241 1.453l-13.001 82.152a1.257 1.257 0 0 1-1.242 1.061H35.113a1.257 1.257 0 0 1-1.234-1.494l13.044-68.085a1.257 1.257 0 0 1 1.234-1.021h35.059c.322 0 .633-.124.866-.346L98.133 66.68c.234-.222.544-.346.867-.346z' fill='%23B3D2FF'/%3E%3Cpath d='M9.055 33.373 3.47 30.48a2.01 2.01 0 0 1 0-3.57l5.586-2.891a2.01 2.01 0 0 0 .863-.865l2.88-5.588a2.01 2.01 0 0 1 3.573 0l2.88 5.588c.19.37.492.673.862.865L25.7 26.91a2.01 2.01 0 0 1 0 3.57l-5.586 2.892a2.01 2.01 0 0 0-.862.864l-2.88 5.588a2.01 2.01 0 0 1-3.573 0l-2.88-5.588a2.01 2.01 0 0 0-.863-.864zM176.237 132.277l-4.615-2.388a1.005 1.005 0 0 1 0-1.785l4.615-2.389c.185-.095.336-.247.431-.432l2.38-4.62a1.005 1.005 0 0 1 1.787 0l2.38 4.62c.096.185.247.337.432.432l4.614 2.389a1.005 1.005 0 0 1 0 1.785l-4.614 2.388a1.005 1.005 0 0 0-.431.433l-2.38 4.62a1.005 1.005 0 0 1-1.788 0l-2.38-4.62a1.005 1.005 0 0 0-.431-.433z' fill='%23E5F0FF'/%3E%3C/g%3E%3C/g%3E%3C/svg%3E")`,
      backgroundRepeat: 'no-repeat',
      backgroundPositionX: 'center',
      backgroundPositionY: 'calc(50% - 100px)',
      color: '#7c87a8',
      textAlign: 'center',
    }}
  >
    <h2 style={{ paddingRight: '20px' }}>No Data</h2>
  </div>
);
