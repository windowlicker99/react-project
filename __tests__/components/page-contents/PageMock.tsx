import React from 'react';
import { render } from '@testing-library/react';
import { PageMock } from '@components/page-contents/PageMock';

describe('PageMock tests', () => {
  it('Should render', () => {
    const { getByText } = render(<PageMock />);
    const text = getByText(/No Data/i);
    expect(text).toBeTruthy();
  });

  it('Should match snapshot', () => {
    const { asFragment } = render(<PageMock />);
    expect(asFragment()).toMatchSnapshot();
  });
});
