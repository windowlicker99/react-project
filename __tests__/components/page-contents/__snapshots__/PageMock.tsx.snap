// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageMock tests Should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; height: 100%; display: flex; justify-content: center; align-items: center; background-repeat: no-repeat; background-position-x: center; background-position-y: calc(50% - 100px); color: rgb(124, 135, 168); text-align: center;"
  >
    <h2
      style="padding-right: 20px;"
    >
      No Data
    </h2>
  </div>
</DocumentFragment>
`;
